nox/manifest.py

```diff
@@ -0,0 +1,166 @@
+# Copyright 2017 Jon Wayne Parrott
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#    http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+from __future__ import absolute_import
+import copy
+import itertools
+
+from nox._parametrize import generate_calls
+from nox.logger import logger
+from nox.sessions import Session
+
+
+class Manifest(object):
+    """Session manifest.
+
+    The session manifest provides the final list of sessions that should be
+    run by nox.
+
+    It is possible for this to be mutated during execution. This allows for
+    useful use cases, such as for one session to "notify" another.
+
+    Args:
+        session_functions (Mapping[str, function]): The registry of discovered
+            session functions.
+        global_config (): The global configuration.
+    """
+    def __init__(self, session_functions, global_config):
+        self._all_sessions = []
+        self._config = global_config
+
+        # Create the sessions based on the provided session functions.
+        for name, func in session_functions.items():
+            self.make_session(name, func)
+
+        # By default, every available session starts off in the queue.
+        self._queue = copy.copy(self._all_sessions)
+        self._consumed = []
+
+        # If the configuration specified filters, apply those now.
+        if self._config.sessions:
+            self.filter_by_name(self._config.sessions)
+        if self._config.keywords:
+            self.filter_by_keywords(self._config.keywords)
+
+    def __iter__(self):
+        return self
+
+    def __next__(self):
+        """Return the next item in the queue.
+
+        Raises:
+            StopIteration: If the queue has been entirely consumed.
+        """
+        if not len(self._queue):
```

Comment:
pylint actually yells about this. I don't feel strongly, though (and I don't use pylint on nox).